### DISCUSSION TITLE
fields:suffix to get standard or custom fields

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#5510] [Ideas]
I want to post this request as an issue so bouncing it here first:

Every now and then I want to list all the standard fields or the custom fields. Currently, I must hard code something like `[...fields[]] -title -tags -created -creator -..... ` or I can be clever with e.g `$vars standardfields={{{ [[dummytiddler]fields[]] }}}` ...but this is all a bit silly.

I propose some suffix solution like so `fields:standard` and `fields:custom`. 

Or is there some existing elegant solution?

**Comment by pmario** [comment]
see: https://tiddlywiki.com/#FieldsWidget:FieldsWidget%20%5B%5Bfields%20Operator%5D%5D%20%5B%5Bfields%20Operator%20(Examples)%5D%5D

**Comment by pmario** [comment]
The only field, that is needed to qualify a tiddler as a tiddler is the `title` field. The rest of the fields is convention only.
 
Conventions are: https://tiddlywiki.com/#TiddlerFields

Where we could already discuss where `caption, list` belong to

**Comment by twMat** [comment]
Thanks @pmario for useful answer. So it appears there is no elegant solution; even `fields:exclude[list title text]]` requires manual typing of what specific fields to exclude.

>The only field, that is needed to qualify a tiddler as a tiddler is the title field. 

He he, that is one poor son-of-a-tiddler.

The docs define (or at least "refers") to some specific fields as "standard fields" [here](https://tiddlywiki.com/#TiddlerFields). My typical use case is that I want to exclude the "non-custom fields" from tiddlers I've created myself, so it is limited to these standard fields. Would be convenient with something like `fields:exclude:standard[]`. I would think others also encounter situations where they want to list only their custom fields in their user created tiddlers.

**Comment by Jermolene** [comment]
Hi @twMat it's a reasonable request, but the trouble is that the definition of "standard fields" might not be consistent across all situations. From the perspective of the core, the standard fields would be title, modified, created, color, tags and list because those are the ones that have special behaviour at the lowest level of the core (there are special setters and getters that handle converting the format of the field). Fields like creator and modifier have some special behaviour but at a somewhat higher level. Then components like $:/Import use yet more fields, which might be considered standard by some.

On the whole, I think the best approach might be to use an explicit definition where it is needed to make things really clear:

```
\define standard-fields() title modified modifier created creator color tags list
```

**Reply by twMat** [reply]
Thanks for reply. Yeah, there is no strict "standard" but the context here is just an available list for end user convenience. 

>On the whole, I think the best approach might be to use an explicit definition [...]

Good - but so I'm requesting that this is a native tiddler. It just isn't worth the effort for the user to have to create such a definition himself at the moment that he needs those items in a filter (besides he is unlikely to look it up here) - so it has to be "on the dinner table already" so to say. Is that OK? It is very small.  (A clear yes or or no is appreciated so the proposal can be closed.)

BTW, I'm surprised the `text` field isn't part of that list?

**Reply by Jermolene** [reply]
> Good - but so I'm requesting that this is a native tiddler. It just isn't worth the effort for the user to have to create such a definition himself at the moment that he needs those items in a filter (besides he is unlikely to look it up here) - so it has to be "on the dinner table already" so to say. Is that OK? It is very small. (A clear yes or or no is appreciated so the proposal can be closed.)

To do that we'd have to decide what constituted the list of "standard" fields, and my point is that there is no objective list of standard fields.

> BTW, I'm surprised the `text` field isn't part of that list?

I missed it! Illustrates the problem rather. Different fields are "standard" at various layers of the design, that's exactly why it's so hard to pin down a single authoritative list that will work for everyone in all situations.

**Reply by twMat** [reply]
>To do that we'd have to decide what constituted the list of "standard" fields, and my point is that there is no objective list of standard fields.

It is only to save the user from having to type the items manually and to remember or investigate which items it is to avoid the very mistake you just made. It doesn't at all have to be an "objective list" for "all situations", it would be a *supporting* list, for many situations, and the user can see what is included in it. It is substantially easier to "confirm" than to "come up with".

**Reply by Jermolene** [reply]
Hi @twMat so what would be the list you would suggest?

**Reply by twMat** [reply]
The fields that are created in a "default tiddler" when the user click the "New tiddler" button and saves. If we looked at all user created tiddlers in existance I would guess these typically have this "set of fields". If the user then wants to add one or two "special" field names to his filter, that is simple and he knows about them because he has likely added those very fields himself to his tiddlers.

Thanks for asking.

**Comment by pmario** [comment]
If you create a new tiddler and then read it from the store with: `$tw.wiki.getTiddler("name")` you get

```
fields: {…}
  ​​created: Date Sat Feb 27 2021 13:11:38 GMT+0100 (Central European Standard Time)
​​  modified: Date Sat Feb 27 2021 13:11:40 GMT+0100 (Central European Standard Time)
​​  tags: Array []
​​  text: ""
​​  title: "New Tiddler"
```

So `creator` and `modifier` are only there if you did define them in the control panel. 

`list` is only there, if the tiddler is a tag _and_ the tags have been sorted by the user.

**Comment by pmario** [comment]
I'm not really sure, what you want to achieve, with a default list. IMO it will always be different, depending on the use-case.

**Reply by twMat** [reply]
I typically want to create a list of the custom fields for tiddlers. To get such a list, I must eliminate the standard fields. There may be *more* fields I need to eliminate, but it would be convenient if one could just easily add all the standard fields in one go. It is *simple* to add standard fields manually, but it is very error prone and requires annoying "thinking" ("...ok, so there's the *title* field... and there's *tags*... and *text*... and isn't is something with the the date, whatsitcalledagain, a *created*... OK, that should be it... Run... Doh! Must add *modified* and *creator*"). Just annoying, that's all.

E.g `list`,`color` and other fields can be included as well in this proposed system tiddler, that's fine. The user can peek at it if he is curious and then edit it if he has special opinions on it. It is just a convenient reference.

**Comment by pmario** [comment]
With 5.1.22 I did add some new parameters to the widget and filters. see https://tiddlywiki.com/#FieldsWidget:FieldsWidget%20%5B%5Bfields%20Operator%20(Examples)%5D%5D%20%5B%5Bfields%20Operator%5D%5D

The include param will also take care, that the order will be preserved. So you should be able to create tiddlers, that contain your list. We probably should add some more examples to the docs.
